mirrord-layer/src/file.rs

```diff
@@ -0,0 +1,100 @@
+use std::{collections::HashMap, env, lazy::SyncLazy, os::unix::io::RawFd, sync::Mutex};
+
+use libc::{c_int, O_ACCMODE, O_APPEND, O_CREAT, O_RDONLY, O_RDWR, O_TRUNC, O_WRONLY};
+use mirrord_protocol::OpenOptionsInternal;
+use regex::RegexSet;
+use tracing::warn;
+
+pub(crate) mod hooks;
+pub(crate) mod ops;
+
+/// Regex that ignores system files + files in the current working directory.
+static IGNORE_FILES: SyncLazy<RegexSet> = SyncLazy::new(|| {
+    // To handle the problem of injecting `open` and friends into project runners (like in a call to
+    // `node app.js`, or `cargo run app`), we're ignoring files from the current working directory.
+    let current_dir = env::current_dir().unwrap();
+
+    let set = RegexSet::new(&[
+        r".*\.so",
+        r".*\.d",
+        r"^/proc/.*",
+        r"^/sys/.*",
+        r"^/lib/.*",
+        r"^/etc/.*",
+        r"^/usr/.*",
+        r"^/dev/.*",
+        r"^/home/iojs/.*",
+        // TODO: `node` searches for this file in multiple directories, bypassing some of our
+        // ignore regexes, maybe other "project runners" will do the same.
+        r".*/package.json",
+        &current_dir.to_string_lossy(),
+    ])
+    .unwrap();
+
+    set
+});
+
+type LocalFd = RawFd;
+type RemoteFd = RawFd;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct RemoteFile {
+    fd: RawFd,
+}
+
+pub(crate) static OPEN_FILES: SyncLazy<Mutex<HashMap<LocalFd, RemoteFd>>> =
+    SyncLazy::new(|| Mutex::new(HashMap::with_capacity(4)));
+
+pub(crate) trait OpenOptionsInternalExt {
+    fn from_flags(flags: c_int) -> Self;
+    fn from_mode(mode: String) -> Self;
+}
+
+impl OpenOptionsInternalExt for OpenOptionsInternal {
+    fn from_flags(flags: c_int) -> Self {
```

Comment:
Is Ext external or extension? Why is this separated from OpenOptionsInternal?

Reply:
`Ext` is a convention for [extension traits](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html).

I could implement `TryFrom<c_int> for OpenOptionsInternal`, and `TryFrom<String> for OpenOptionsInternal`, but I thought such conversions did not belong in the `mirrord-protocol` crate (they're conversions that will only ever be used by `mirrord-layer` as far as I can see). Plus adding these conversions in the `-protocol` crate would require adding `libc` as a dependency, to access the libc flags in `from_flags`.